### PR TITLE
Unique token improvement

### PR DESCRIPTION
**Pull request by crtrott**
Align interface with other Kokkos interfaces which default the execution space. 

This also removes the usage of bitset in favor of using straight up int arrays as locks for UniqueToken in HIP and CUDA. 
As a consequence atomic conflicts are vastly reduced, and also by default you are getting much better aligned indicies within a warp, improving data access patterns in common usecases. 

Here is a performance example:
```c++
#include<Kokkos_Core.hpp>

int main(int argc, char* argv[]) {
  Kokkos::initialize(argc,argv);
  {
     int N = (argc>1) ? atoi(argv[1]) : 30000;
     int R = (argc>2) ? atoi(argv[2]) : 1;


     Kokkos::Experimental::UniqueToken<> token;
     Kokkos::View<double*> a("A",token.size());

     auto f1 = KOKKOS_LAMBDA(int) {
       int idx = token.acquire();
       for(int r=0; r<R; r++)
         a(idx) += a(idx)+1;
       Kokkos::memory_fence();
       token.release(idx);
     };
     auto f2 = KOKKOS_LAMBDA(int i) {
       int idx = i % a.extent(0);
       for(int r=0; r<R; r++)
         a(idx) += a(idx)+1;
       Kokkos::memory_fence();
     };

     Kokkos::parallel_for(N, f1);
     Kokkos::parallel_for(N, f2);
     Kokkos::fence();

     Kokkos::Timer timer;
     Kokkos::parallel_for(N, f1);
     Kokkos::fence();
     double time1 = timer.seconds();
     timer.reset();
     Kokkos::parallel_for(N, f2);
     Kokkos::fence();
     double time2 = timer.seconds();

     printf("%lf %lf\n",time1*1.e6, time2*1.e6);
  }
  Kokkos::finalize();
}
```

Prior to the change:
A100: 
```
bash-4.2$ ./test.cuda 30000 1
1823.098000 16.880000
bash-4.2$ ./test.cuda 30000000 1
152508.359000 207.732000
```
MI100:
```
// MI100
bash-4.2$ ./test.host 30000 1
1585.862000 25.080000
bash-4.2$ ./test.host 30000000 1
495600.250000 357.516000
```

Post change:
A100
```
bash-4.2$ ./test.cuda 30000 1
17.940000 14.940000
bash-4.2$ ./test.cuda 30000000 1
836.698000 370.924000
```

MI100
```
bash-4.2$ ./test.host 30000 1
26.680000 24.060000
bash-4.2$ ./test.host 30000000 1
958.009000 353.497000
```

### CODE REVIEW

File: core/src/Cuda/Kokkos_Cuda_UniqueToken.hpp [hunk 1]
```diff
@@ -88,22 +87,54 @@ class UniqueToken<Cuda, UniqueTokenScope::Global> {
...
   KOKKOS_INLINE_FUNCTION
   size_type acquire() const {
-    const Kokkos::pair<int, int> result =
-        Kokkos::Impl::concurrent_bitset::acquire_bounded(
-            m_buffer, m_count, Kokkos::Impl::clock_tic() % m_count);
-
-    if (result.first < 0) {
-      Kokkos::abort(
-          "UniqueToken<Cuda> failure to acquire tokens, no tokens available");
+#ifdef __CUDA_ARCH__
```
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_UniqueToken.hpp:90]: use `KOKKOS_IF_ON_DEVICE`

File: core/src/Cuda/Kokkos_Cuda_UniqueToken.hpp [hunk 2]
```diff
@@ -117,12 +148,10 @@ class UniqueToken<Cuda, UniqueTokenScope::Instance>
...
-  UniqueToken(size_type max_size, execution_space const& = execution_space())
-      : m_buffer_view(
-            "UniqueToken::m_buffer_view",
-            ::Kokkos::Impl::concurrent_bitset::buffer_bound(max_size)) {
-    m_buffer = m_buffer_view.data();
-    m_count  = max_size;
+  UniqueToken(size_type max_size, execution_space const& = execution_space()) {
+    m_locks = Kokkos::View<uint32_t*, Kokkos::CudaSpace>(
+        "Kokkos::UniqueToken::m_locks", max_size);
+    m_count = max_size;
```
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_UniqueToken.hpp:154]:
```suggestion
  UniqueToken(size_type max_size, execution_space const& = execution_space())
      : m_locks{"Kokkos::UniqueToken::m_locks", max_size}
      , m_count{max_size} {
```

Comment (dalg24) [reply core/src/Cuda/Kokkos_Cuda_UniqueToken.hpp:154]: Note that we could get rid of `m_count` since it is always `m_locks.size()`.

File: core/src/Cuda/Kokkos_Cuda_UniqueToken.hpp [hunk 3]
```diff
@@ -88,22 +87,54 @@ class UniqueToken<Cuda, UniqueTokenScope::Global> {
...
+#endif
   }
 
   /// \brief release an acquired value
   KOKKOS_INLINE_FUNCTION
-  void release(size_type i) const noexcept {
-    Kokkos::Impl::concurrent_bitset::release(m_buffer, i);
+  void release(size_type idx) const noexcept {
+#ifdef KOKKOS_ENABLE_IMPL_DESUL_ATOMICS
+    desul::atomic_thread_fence(desul::MemoryOrderRelease(),
```
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_UniqueToken.hpp:132]: Don't you need an acquire?

**Review by dalg24 — CHANGES_REQUESTED** [review]

File: core/src/Cuda/Kokkos_Cuda_UniqueToken.hpp (hunk 1 above)
Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_UniqueToken.hpp]: done

File: core/src/Cuda/Kokkos_Cuda_Instance.cpp [hunk 4]
```diff
@@ -1003,8 +1003,9 @@ namespace Experimental {
 
 UniqueToken<Kokkos::Cuda, Kokkos::Experimental::UniqueTokenScope::Global>::
     UniqueToken(Kokkos::Cuda const &)
-    : m_buffer(
-          Kokkos::Impl::CudaInternal::singleton().m_scratchConcurrentBitset),
+    : m_locks(Kokkos::View<uint32_t *, Kokkos::CudaSpace>(
+          "Kokkos::UniqueToken::m_locks",
+          Kokkos::Impl::CudaInternal::singleton().m_maxConcurrency)),
```
[the remark this answers is not in the document]
Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_Instance.cpp]: done

File: core/src/HIP/Kokkos_HIP_UniqueToken.hpp [hunk 5]
```diff
@@ -45,9 +45,12 @@
 #ifndef KOKKOS_HIP_UNIQUE_TOKEN_HPP
 #define KOKKOS_HIP_UNIQUE_TOKEN_HPP
 
-#include <impl/Kokkos_ConcurrentBitset.hpp>
+#include <Kokkos_Macros.hpp>
+#ifdef KOKKOS_ENABLE_HIP
```
Comment (Rombur) [inline core/src/HIP/Kokkos_HIP_UniqueToken.hpp:49]: This is not needed. If HIP is not enabled, CMake will ignore the file

File: core/src/HIP/Kokkos_HIP_UniqueToken.hpp [hunk 6]
```diff
@@ -81,49 +83,68 @@ class UniqueToken<HIP, UniqueTokenScope::Global> {
...
+            } else {
+              idx += blockDim.y * blockDim.x + 1;
+              idx = idx % size();
+            }
+          }
+          done_active = __ballot(done ? 1 : 0);
+        } return idx;
+
+// Make sure that all writes in the previous lock owner are visible to me
+#ifdef KOKKOS_ENABLE_IMPL_DESUL_ATOMICS
```
Comment (Rombur) [inline core/src/HIP/Kokkos_HIP_UniqueToken.hpp:109]: The compiler does not like the macro inside another macro
`/var/jenkins/workspace/Kokkos/core/src/HIP/Kokkos_HIP_UniqueToken.hpp:109:2: error: embedding a directive within macro arguments has undefined behavior [-Werror,-Wembedded-directive]
`

Comment (dalg24) [reply core/src/HIP/Kokkos_HIP_UniqueToken.hpp:109]: Use an `impl_acquire` function like you did in Cuda.

File: core/src/Kokkos_UniqueToken.hpp [hunk 7]
```diff
@@ -59,8 +59,8 @@ enum class UniqueTokenScope : int { Instance, Global };
 ///
 /// This object should behave like a ref-counted object, so that when the last
 /// instance is destroy resources are free if needed
-template <typename ExecutionSpace,
-          UniqueTokenScope = UniqueTokenScope::Instance>
+template <typename ExecutionSpace = Kokkos::DefaultExecutionSpace,
```
Comment (dalg24) [inline core/src/Kokkos_UniqueToken.hpp:62]: Why did you change?  This is not in the description of the PR.

File: core/src/Cuda/Kokkos_Cuda_UniqueToken.hpp [hunk 8]
```diff
@@ -83,46 +83,79 @@ class UniqueToken<Cuda, UniqueTokenScope::Global> {
...
+#else
+    while (Kokkos::atomic_compare_exchange(&m_locks(idx), 0, 1) == 1) {
+      idx += blockDim.y * blockDim.x + 1;
+      idx = idx % size();
     }
+#endif
+// Make sure that all writes in the previous lock owner are visible to me
+#ifdef KOKKOS_ENABLE_IMPL_DESUL_ATOMICS
+    desul::atomic_thread_fence(desul::MemoryOrderAcquire(),
+                               desul::MemoryScopeDevice());
```
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_UniqueToken.hpp:120]: What writes?  To `&m_locks(idx)`?  If so is it really needed.  It looks like all accesses to `m_lock` are atomic operations.

Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_UniqueToken.hpp:120]: writes happening between acquire and release of the token ??

File: core/src/Kokkos_UniqueToken.hpp (hunk 7 above)
Comment (crtrott) [reply core/src/Kokkos_UniqueToken.hpp:62]: It is now ...

Everything else in Kokkos defaults the execution space. I think this should too.

File: core/src/HIP/Kokkos_HIP_UniqueToken.hpp (hunk 6 above)
Comment (crtrott) [reply core/src/HIP/Kokkos_HIP_UniqueToken.hpp:110]: done

File: core/src/HIP/Kokkos_HIP_UniqueToken.hpp (hunk 5 above)
Comment (crtrott) [reply core/src/HIP/Kokkos_HIP_UniqueToken.hpp]: removed

File: core/src/Cuda/Kokkos_Cuda_UniqueToken.hpp [hunk 9]
```diff
@@ -60,14 +59,15 @@ namespace Experimental {
...
+  Kokkos::View<uint32_t*, Kokkos::CudaSpace> m_locks;
 
  public:
   using execution_space = Cuda;
   using size_type       = int32_t;
 
-  explicit UniqueToken(execution_space const& = execution_space());
+  explicit UniqueToken(execution_space const& = execution_space())
+      : m_locks(Kokkos::View<uint32_t*, Kokkos::CudaSpace>(
+            "Kokkos::UniqueToken::m_locks", Kokkos::Cuda().concurrency())) {}
```
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_UniqueToken.hpp:70]: Use the exec space provided to initialize your locks

**Comment by ajpowelsnl** [comment]
Hi @crtrott and @dalg24 a couple of questions about this issue:  1) will this item be in Kokkos 3.6?  2) if so, is it still blocked?

**Comment by crtrott** [comment]
@dalg24 @masterleinad finally passed!!

File: core/src/Cuda/Kokkos_Cuda_UniqueToken.hpp [hunk 10]
```diff
@@ -60,14 +59,19 @@ namespace Experimental {
...
+  Kokkos::View<uint32_t*, Kokkos::CudaSpace> m_locks;
 
  public:
   using execution_space = Cuda;
   using size_type       = int32_t;
 
-  explicit UniqueToken(execution_space const& = execution_space());
+  explicit UniqueToken()
+      : m_locks(Kokkos::View<uint32_t*, Kokkos::CudaSpace>(
+            "Kokkos::UniqueToken::m_locks", Kokkos::Cuda().concurrency())) {}
```
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_UniqueToken.hpp:70]:
```suggestion
      : m_locks("Kokkos::UniqueToken::m_locks", Kokkos::Cuda().concurrency()) {}
```
haven't checked the format

File: core/src/Cuda/Kokkos_Cuda_UniqueToken.hpp [hunk 11]
```diff
@@ -60,14 +59,19 @@ namespace Experimental {
...
+  explicit UniqueToken(execution_space const& exec)
+      : m_locks(Kokkos::View<uint32_t*, Kokkos::CudaSpace>(
+            Kokkos::view_alloc(exec, "Kokkos::UniqueToken::m_locks"),
+            Kokkos::Cuda().concurrency())) {}
```
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_UniqueToken.hpp:74]:
```suggestion
      : m_locks(Kokkos::view_alloc(exec, "Kokkos::UniqueToken::m_locks"),
                Kokkos::Cuda().concurrency()) {}
```
not formatted

File: core/src/Cuda/Kokkos_Cuda_UniqueToken.hpp [hunk 12]
```diff
@@ -83,46 +87,84 @@ class UniqueToken<Cuda, UniqueTokenScope::Global> {
...
-            ::Kokkos::Impl::concurrent_bitset::buffer_bound(max_size)) {
-    m_buffer = m_buffer_view.data();
-    m_count  = max_size;
+  UniqueToken(size_type max_size) {
+    m_locks = Kokkos::View<uint32_t*, Kokkos::CudaSpace>(
+        "Kokkos::UniqueToken::m_locks", max_size);
+  }
+  UniqueToken(size_type max_size, execution_space const& exec) {
+    m_locks = Kokkos::View<uint32_t*, Kokkos::CudaSpace>(
+        Kokkos::view_alloc(exec, "Kokkos::UniqueToken::m_locks"), max_size);
```
Comment (masterleinad) [inline core/src/Cuda/Kokkos_Cuda_UniqueToken.hpp:167]: I slightly prefer to define and call protected constructors here to avoid allocating memory for `m_locks` twice like in https://github.com/kokkos/kokkos/pull/4748/files.

File: core/src/HIP/Kokkos_HIP_UniqueToken.hpp [hunk 13]
```diff
@@ -45,9 +45,11 @@
 #ifndef KOKKOS_HIP_UNIQUE_TOKEN_HPP
 #define KOKKOS_HIP_UNIQUE_TOKEN_HPP
 
-#include <impl/Kokkos_ConcurrentBitset.hpp>
+#include <Kokkos_Macros.hpp>
+
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_UniqueToken.hpp:49]:
```suggestion
```

File: core/src/Cuda/Kokkos_Cuda_UniqueToken.hpp (hunk 12 above)
Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_UniqueToken.hpp:167]: what do you mean allocating twice?

**Review by dalg24 — APPROVED** [review]

**Comment by masterleinad** [comment]
It looks like `m_scratchConcurrentBitset` is unused now.

**Comment by masterleinad** [comment]
We should still be able to get rid of the now unused `m_scratchConcurrentBitset`.

**Comment by dalg24** [comment]
Retest this please

**Comment by dalg24** [comment]
```
4: [ RUN      ] openmptarget.unique_token_global
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestUniqueToken.hpp:151: Failure
4: Expected equality of these values:
4:   sum
4:     Which is: 9999980
4:   int64_t(N) * R
4:     Which is: 10000000
4: [  FAILED  ] openmptarget.unique_token_global (280 ms)

**Comment by crtrott** [comment]
I don't get how either of these guys is failing (OpenMPTarget and OpenMP for NVHPC), I didn't really touch that stuff at all did I?

**Comment by dalg24** [comment]
Retest this please

**Review by masterleinad — CHANGES_REQUESTED** [review]
Why should we keep `m_scratchConcurrentBitset`?

**Review by masterleinad — APPROVED** [review]

**Comment by dalg24** [comment]
Failure is unrelated (Clang crashing when compiling Kokkos::complex unit test, sigh)
